content/blog/2022-03-08-alive/index.md

```diff
@@ -0,0 +1,114 @@
+
+### Undefined Behaviors in C/C++
+Finally, it is worth understanding that C and C++ both have a variety of ways to write code which result in  _undefined behavior_, where the specification of the language says that anything is allowed to happen. For example, if you try to index an array outside of its bounds, it is okay for the program to crash, but it is also okay for it to return data, print something out, or literally anything that it wants.
+
+Given the prevalence of security vulnerabilities caused by improper bounds checking in array accesses such as [Heartbleed](https://en.wikipedia.org/wiki/Heartbleed), why do C specifications let you have undefined behavior rather than just defining that the behavior should be a program crash?
+
+Ralf Jung has [a good discussion](https://blog.sigplan.org/2021/11/18/undefined-behavior-deserves-a-better-reputation/), but there was also a simple (and reasonably convincing to me) example from the class discussion. When people are writing in C, they are much more likely to care a lot about performance. If the programmer knows that they didn’t mess up their loop to iterate through all of the elements of the array, then doing bounds-checking on every array access will slow down your program unnecessarily, and doing bounds checking only when it might be useful might be too hard for your compiler to do.
+
+In a similar flavor, undefined behavior allows a compiler to more aggressively optimize code, since while the optimized code must generate the same results whenever the results are defined, it is free to do optimizations and take shortcuts which further modify (mangle?) the code whenever it would produce an undefined result.[^1]
+
+[^1]: If this is unsatisfying, you can also wait until later, where we will talk about this again.
+
+## Main Contributions
+With the background out of the way, I reiterate:
+
+> At its core, Alive is a domain specific language for formally verifying peephole optimizations in LLVM.  Empirically, it found bugs where the optimization could introduce additional undefined behavior in situations where the original code did not.
+
+### Formal Verification of Peephole Optimizations
+A core engine of the paper is in creating a domain specific language which makes it easy to write peephole optimizations which can be formally verified.
+
+#### Basic Overview
+An LLVM instruction can have source preconditions $\phi$ and target postconditions $\bar \phi$, source definedness constraints $\delta$, a source poison-free constraint $\rho$, a source result $\iota$, as well as target definedness, poison-free constraint, and result  $\bar delta$, $\bar \rho$, andt $\bar \iota$ respectively.
+
+*Preconditions* can be defined using a must-analysis for precise constraints, and may-analysis for over/underapproximations. For example, the precise `isPowerOfTwo(%a)` constraint is defined as $p \leftarrow a \neq 0 \wedge a &(a-1) = 0$, while the approximate `mayAlias(%a, %b)` 
```

Comment:
Not quite sure what you mean by the $\leftarrow$ symbol… is this implication? If so, maybe you want a rightward-facing arrow of some sort. It also looks like the sentence got cut off here.